components/esp_dns/esp_dns_utils.c

```diff
@@ -0,0 +1,211 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <arpa/inet.h>
+#include "esp_random.h"
+#include "esp_dns_utils.h"
+
+/* Function to create a DNS query */
+size_t create_dns_query(uint8_t *buffer, size_t buffer_size, const char *hostname, int addrtype, uint16_t *id_o)
+{
+    /*
+    Sample DNS Query for example.com (Type A)
+    0x00, 0x00,       // Transaction ID
+    0x01, 0x00,       // Flags: Standard query
+    0x00, 0x01,       // Questions: 1
+    0x00, 0x00,       // Answer RRs: 0
+    0x00, 0x00,       // Authority RRs: 0
+    0x00, 0x00,       // Additional RRs: 0
+    0x07, 'e', 'x', 'a', 'm', 'p', 'l', 'e', // QNAME: example.com
+    0x03, 'c', 'o', 'm',
+    0x00,             // End of QNAME
+    0x00, 0x01,       // QTYPE: A (host address)
+    0x00, 0x01        // QCLASS: IN (internet)
+    */
+
+    dns_header_t *header = (dns_header_t *)buffer;
+    memset(buffer, 0, buffer_size);
+
+    /* Set header fields */
+    *id_o = (uint16_t)(esp_random() & 0xFFFF); /* Return the id for response validation */
+    header->id = htons(*id_o);         /* Random transaction ID */
+    header->flags = htons(0x0100);      /* Standard query with recursion */
+    header->qdcount = htons(1);         /* One question */
+
+    /* Add the question name */
+    uint8_t *qname = buffer + sizeof(dns_header_t);
+    const char *dot = hostname;
+    while (*dot) {
+        const char *next_dot = strchr(dot, '.');
+        if (!next_dot) {
+            next_dot = dot + strlen(dot);
+        }
+        uint8_t len = next_dot - dot;
+        *qname++ = len;
+        /* Check for overflow */
+        if ((qname - buffer) > buffer_size) {
+            return -1;
+        }
+        memcpy(qname, dot, len);
+        qname += len;
+        dot = (*next_dot) ? next_dot + 1 : next_dot;
+    }
+    *qname++ = 0;  /* Null-terminate the question name */
+
+    /* Set question fields */
+    dns_question_t *question = (dns_question_t *)qname;
+    question->qtype = htons(addrtype);
+    question->qclass = htons(DNS_RRCLASS_IN);
+
+    /* Return the total query size */
+    return (qname + sizeof(dns_question_t)) - buffer;
+}
+
+/**
+ * Skips over a DNS name in a DNS reply message and returns the offset to the end of the name.
+ *
+ * This function handles both uncompressed labels and compression pointers according to RFC 1035.
+ * Reference: RFC 1035, sections 3.1 (Name Space Definitions) and 4.1.4 (Message Compression).
+ *
+ * @param ptr Pointer to the start of the DNS name in the DNS message.
+ * @return The ptr to the end of the DNS name.
+ */
+static uint8_t *skip_dns_name(uint8_t *ptr)
+{
+    uint8_t offset = 0;
+
+    /* Loop through each part of the name, handling labels and compression pointers */
+    while (ptr[offset] != 0) {
```

Comment:
should we also check if wer'e not past the parsed data? (you can receive anything from the network, e.g. a malicious packet)?

Reply:
Added a check to sanitize offset.